### PR TITLE
Per-timestep EventProp loss

### PR DESCRIPTION
**Pull request by neworderofjamie**
As promised, this PR re-adds per-timestep loss to EventProp. Because it adds a chain of softmax custom updates every timestep, it is significantly slower than the per-batch loss functions.

Note, this requires https://github.com/genn-team/genn/tree/origin/egp_ref which I will also make a PR for.

### CODE REVIEW

**Review by tnowotny — APPROVED** [review]
Loks good overall though a difficult one to parse. I did not check details fo indexing into the ring buffers but would trust that if an example (MNIST?) appears to run fine (expect good accuracy on MNIST).

**Comment by neworderofjamie** [comment]
Just verified everything is working on my machine and obtained the following after 10 epochs on my machine:
||Train accuracy [%]|Test accuracy [%]|Train time[s]|
|-|-|-|-|
|Per-timestep|96.35|96.6|50
|Per-trial|96.5|96.4|28
